sdk/data/azcosmos/partition_key_builder.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+// PartitionKeyBuilder is used to build a PartitionKey.
+type PartitionKeyBuilder struct {
```

Comment:
My gut feeling is that we don't need a brand-new builder type here. The existing `PartitionKey` struct already holds a `values []interface{}` value, and these `Append` methods could just be added to that type, making it mutable and supporting syntax like this:

```go
pk := NewPartitionKeyString("key0").AppendString("key1")
```

Now, I'll confess to not being quite as up-to-speed on the latest Go idioms, so I'm not sure how strongly we feel about making a currently-immutable type mutable, but putting the `Append` methods on `PartitionKey` feels more compact to me. It also removes the ambiguity of having both a "Builder" API for partition keys and direct constructors.

Reply:
The main reason behind this design is that the same pattern is found in [azure-cosmos-dotnet-v3](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5e6232d25f84074d5008e97670d3e0dc07891f0e/Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs), but I'll gladly refactor the PR to this design instead.

An alternative could be `func NewPartitionKey(parts ...interface{})`.

Reply:
> The main reason behind this design is that the same pattern is found in [azure-cosmos-dotnet-v3](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5e6232d25f84074d5008e97670d3e0dc07891f0e/Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs), but I'll gladly refactor the PR to this design instead.

Gotcha. I don't have direct context on why .NET does it this way, but the main thing it does is some validation, which we've generally tried to leave to the server where possible (so that the client doesn't end up restricting something that is later allowed.

Not having the full context, if I were designing the .NET API at this point I think I'd suggest a similar design to what I'm suggesting here in Go, so I'm not too worried if we diverge from how the .NET SDK does things here. We can see what other folks on the SDK team think about that too.

> An alternative could be func NewPartitionKey(parts ...interface{}).

Yeah, I considered that, but we lose out on type-checking, since you could put anything in there. We'd have to either return an error out of `NewPartitionKey` (which means consumers have to add a bunch of code), panic (which is usually something to avoid, though this is a "programming" error rather than a recoverable error), or report that error when the request is serialized (which feels "too late").

I like the `AppendString`/`AppendNumber`/etc. patterns here, so I think I'd like to keep those if possible.

Reply:
@RickWinter - any thoughts on the API pattern and Go idioms here?

Reply:
I too would prefer the APIs on the `PartitionKey` type as having a dedicated builder seems C#/Java-ish. There are a few warts though.
- The current constructors return `PartitionKey` by value so you can't chain the calls (e.g. `NewPartitionKeyString("foo").AppendNumber(123)` won't compile). While we can fix this with a `NewPartitionKey() *PartitionKey` constructor, the lack of uniformity is a little clunky (changing the existing constructors to return pointer-to-type is a breaking change which we should avoid).
- APIs take `PartitionKey` by value so one would have to dereference the key var before/at the call site which looks a little weird (e.g. `SomeAPI(*NewPartitionKey().AppendString("foo"))`).

Having the builder elides the above issues at the cost of introducing a new type and its constructor. So, while I could live with the second bullet, the first one is nudging me towards the builder.

Reply:
Side-by-side comparison.
```go
	SomeAPI(*NewPartitionKey().AppendString("foo"))

	// vs

	SomeAPI(NewPartitionKeyBuilder().AppendString("foo").Build())
```
Thoughts?

Reply:
Oof, yeah I hadn't considered that issue (I'll admit it's been a bit since I worked actively in Go so I forgot it doesn't support auto-referencing to call methods).

Given those restrictions, I think the builder might be the best approach here. I don't love that the builder requires an additional `Build()` call, but I agree that the inconsistent constructors would be strange.

Another option is for the `Append` APIs to take the receiver by value:

```go
func (pk PartitionKey) AppendString(value string) PartitionKey {
	pk.components = append(pkb.components, value)
	return pk
}
```

That would allow this code to continue working:

```go
SomeAPI(NewPartitionKeyString("parent").AppendString("child"))
```

The syntax for appending to a `PartitionKey` stored in a variable isn't as tidy: `pk = pk.AppendString("...")` (though this mirrors things like `append`, and there's possibly _some_ kind of impact on performance (though tbh it could go either way depending on things like copy elision for by-value and memory barriers for by-reference, so we really don't know without measuring).

Reply:
Good idea, and I think I can live with that trade-off assuming that callers will either capture the final `PartitionKey` (`pk := NewPartitionKey().AppendString("foo").AppendString("bar")`) or make the call in-place like above.

Reply:
I'll have a go at the suggestions tomorrow morning (EU time) :^)

Reply:
That's perfect, I noticed the errors before but didn't know whether/how I should address it since it wasn't caused by anything in this PR.